### PR TITLE
Adds an RNG subsystem

### PR DESCRIPTION
**Pull request by MrPerson**
Replaces all rand(), prob(), and pick() calls with SSrng.random(), SSrng.probability(), and SSrng.pick_from_list(). Do not call rand(), prob(), or pick() directly anymore. There's no direct benefits for doing this, just laying the groundwork for future stuff like seeded map generation or w/e.

walk_rand(), get_step_rand(), and roll() would all also increment BYOND's rng tracker and would also need to be tracked, but none of those procs are used in the codebase so it's not necessary. And keep it that way you fuckwits.

Gets rid of all cases of that stupid pick() w/ weights format which I swear half the time people were calling backwards anyway. We have pickweight, just use that.

Because this is going to conflict a lot, I'll resolve conflicts right before it gets merged.

### CODE REVIEW

**Comment by Shadowlight213** [comment]
Wait, does this mean we can have goof's dream of doing a round where there's no rng, and every call to prob returns true?

File: code/controllers/subsystem/rng.dm [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		return
+	else if(_offset < total_offset)
+		set_seed(seed) //total_offset is now 0
+	if(!total_offset && seed == -1)
+		set_seed()
+	while(total_offset < _offset)
+		random()
+
+/datum/controller/subsystem/rng/proc/random(lower, upper) //replaces rand()
+	if(total_offset++ && seed == -1) //only tests the seed once
```
Comment (RemieRichards) [inline code/controllers/subsystem/rng.dm]:
```dm
#define VALIDATE_SEED(procname)                                             \
if(total_offset++ && seed == -1) //only tests the seed once                 \
    total_offset = 0                                                        \
    stack_trace("SSrng.##procname() was called BEFORE a seed was set")      \
```

**Comment by XDTM** [comment]
Finally i won't keep rolling the same 3 symptoms for half an hour in virology

**Comment by RemieRichards** [comment]
No, you will.
This is still byond RNG, we can just say "this round's RNG will be the same as round 237" was (by giving this round, and round 237, the same seed)

**Comment by MrPerson** [comment]
Note that we could always do other stuff like roll our own RNG system if we really wanted to. This might be good if we wanted multiple independent systems. For example if we wanted to generate maint all with a seed of 1 but CHECK_TICK sleeps the proc, by the time it wakes up, it might be on a different offset because other things called some random procs in the meantime. Then if you generate again you'll get a different result with the same seed. If there's one thing that absolutely 100% should be determinant, it's the RNG seed.

**Comment by FantasticFwoosh** [comment]
- Have lucky rabbits foot admin-item = disable RNG subsystem while in inventory.

**Comment by Ratvar** [comment]
> Lucky foot <
Why in inventory? Real lucky foot, cut off leg to enable RNG again. New and cool wizard items?

**Comment by kevinz000** [comment]
>no admin abuse overrides

:'(

File: code/controllers/subsystem/rng.dm [hunk 2]
```diff
@@ -0,0 +1,70 @@
+//If we do rand_seed(1); prob(50); rand(1, 10); rand(1, 100)
+//and then rand_seed(1); rand(); rand(1, 10); rand(1, 100)
+
+//Both rand(1, 10) and rand(1, 100) calls will return the same result, every time
+//This is why we care about the total offset, so we know where we are in the chain
+
+SUBSYSTEM_DEF(rng)
+	name = "RNG"
+	init_order = INIT_ORDER_RNG
+	flags = SS_NO_FIRE
```
Comment (Cyberboss) [inline code/controllers/subsystem/rng.dm]: SS_NO_INIT

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/__DEFINES/subsystems.dm [hunk 3]
```diff
@@ -43,9 +43,10 @@
 // Subsystems shutdown in the reverse of the order they initialize in
 // The numbers just define the ordering, they are meaningless otherwise.
 
-#define INIT_ORDER_DBCORE 18
-#define INIT_ORDER_BLACKBOX 17
-#define INIT_ORDER_SERVER_MAINT 16
+#define INIT_ORDER_DBCORE 19
+#define INIT_ORDER_BLACKBOX 18
+#define INIT_ORDER_SERVER_MAINT 17
+#define INIT_ORDER_RNG 16
```
Comment (Cyberboss) [inline code/__DEFINES/subsystems.dm]: I'm going to need to do an MC change to make PreInit() respect init_order because currently it doesn't

**Comment by ninjanomnom** [comment]
A performance profile of these would be useful compared to the base byond procs.

**Comment by DoctoreMad** [comment]
>There's no direct benefits for doing this, just laying the groundwork for future stuff like seeded map generation or w/e.

**Comment by JJRcop** [comment]
This is one step closer to replayable rounds!

**Comment by Xhuis** [comment]
As a suggestion, perhaps it'd be a good idea to use wrappers, i.e. have a `Prob(X)` or something similarly-named proc that just calls ``SSrng.probability(X)`? Just an idea.

**Comment by Cyberboss** [comment]
It's really just a preference thing at that point, Also, it's impossible to mess up grepping the full call

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Also do what Remie said

**Comment by Fox-McCloud** [comment]
This seems like needless performance overhead, to be honest. If seed tracking was actually implemented and utilized, I could see this potentially being viable (for those *specific* things that need to use it), but blatantly replacing all RNG with its own subsystem seems like needless overcomplexity and a sizeable decrease in performance.

Doesn't really seem, to me that this would be necessary for the vast bulk of RNG systems, and only a very specific subset.

**Comment by Cyberboss** [comment]
I'm not at all worried about performance in this case considering I believe no subsystem uses rng off the top of my head

**Comment by MrPerson** [comment]
You have to manage the offset globally or you might as well not even bother.

If there's a profile of this being really awfully slow then yeah just get rid of it, but frankly we do RNG stuff way less often than you might think relative to a lot of other stuff.

**Comment by Cyberboss** [comment]
K once  more, then I'll merge

**Comment by lordpidey** [comment]
Would it be possible to modify the pickweight() proc while we're already digging around in the RNG methods?

Right now pickweight will sometimes pick 0 weighted things, by setting their weight to 1.  You'll also want to change the last if statement to `if (total <= 0 && L[item])`

**Comment by Tacolizard** [comment]
@tgstation/commit-access

why do we use BYOND RNG? Would it really be that difficult to create our own RNG? And if it's too difficult to create good RNG in BYOND then why don't we use a DLL?

People constantly complain about BYOND RNG when it seems like it would be easy to make our own.

**Comment by RemieRichards** [comment]
@Tacolizard Byond RNG is good enough, people just want a scapegoat

**Comment by Cyberboss** [comment]
@MrPerson pls

**Comment by Cyberboss** [comment]
p l s s s s

**Comment by optimumtact** [comment]
you gotta finish something one day MR p

**Pull request closed without merging**